generator/src/test/java/com/scottlogic/deg/generator/inputs/profileviolation/TypeEqualityHelper.java

```diff
@@ -0,0 +1,106 @@
+            assertRuleTypeEquality(expectedRules.get(i), actualRules.get(i));
+        }
+    }
+
+    /**
+     * Asserts all constraints in two rules have the same type and all sub-constraints of these constraints have the
+     * same type.
+     * @param expectedRule The expected rule.
+     * @param actualRule The actual rule
+     */
+    public static void assertRuleTypeEquality(Rule expectedRule, Rule actualRule) {
+        ArrayList<Constraint> expectedConstraints = new ArrayList<>(expectedRule.constraints);
+        ArrayList<Constraint> actualConstraints = new ArrayList<>(actualRule.constraints);
+        for (int i = 0; i < expectedConstraints.size(); i++) {
+            assertConstraintTypeEquality(expectedConstraints.get(i), actualConstraints.get(i));
+        }
+    }
+
+    /**
+     * Asserts deep type equality for constraints that can contain other constraints.
+     * For example asserts that an AND(OR(X,Y),OR(Z,W)) is not the same as AND(OR(X,Y),AND(Z,W))
+     * but cannot distinguish between constraints A:(foo less than 10) and B:(bar less than 50)
+     * @param expectedConstraint The expected constraint.
+     * @param actualConstraint The constraint under test.
+     */
+    public static void assertConstraintTypeEquality(Constraint expectedConstraint, Constraint actualConstraint) {
+        if (expectedConstraint == null && actualConstraint == null) {
+            return;
+        }
+        Assert.assertEquals("Class types do not match for constraints. Expected: "
+                + expectedConstraint.getClass() + " but was: " + actualConstraint.getClass(),
+            expectedConstraint.getClass(),
+            actualConstraint.getClass());
+        if (expectedConstraint instanceof NotConstraint){
+            assertConstraintTypeEquality(
+                ((NotConstraint)expectedConstraint).negatedConstraint,
+                ((NotConstraint)actualConstraint).negatedConstraint);
+        }
+        else if (expectedConstraint instanceof AndConstraint) {
+            ArrayList<Constraint> expectedConstraints = new ArrayList<>(((AndConstraint) expectedConstraint).subConstraints);
+            ArrayList<Constraint> actualConstraints = new ArrayList<>(((AndConstraint) actualConstraint).subConstraints);
+            for (int i = 0; i < ((AndConstraint) expectedConstraint).subConstraints.size(); i++) {
+                assertConstraintTypeEquality(
+                    expectedConstraints.get(i),
+                    actualConstraints.get(i));
+            }
+        }
+        else if (expectedConstraint instanceof OrConstraint) {
+            ArrayList<Constraint> expectedConstraints = new ArrayList<>(((OrConstraint) expectedConstraint).subConstraints);
+            ArrayList<Constraint> actualConstraints = new ArrayList<>(((OrConstraint) actualConstraint).subConstraints);
+            for (int i = 0; i < ((OrConstraint) expectedConstraint).subConstraints.size(); i++) {
```

Comment:
Assert that both lists are of the same size, otherwise a) it could give a false positive if actual is longer than expected or b) throw an out-of-range exception if actual is shorter than expected